### ISSUE TITLE
Code generation should skip JDBC connection if possible

### ISSUE THREAD

**Issue by seanf** [jOOQ/jOOQ#13263]
### Use case:

I have jOOQ set up to generate code from a Postgres database provided by `org.testcontainers.jdbc.ContainerDatabaseDriver`, which initialises a schema from an SQL file when the container starts up. (The effect is similar to https://www.jooq.org/doc/latest/manual/code-generation/codegen-ddl/ but without the limitations of H2.)

The container startup takes a few seconds, so I would like to skip it (and jOOQ codegen) when the schema file is unchanged. I generate a checksum from the SQL file, and use that checksum as a `schemaVersionProvider`. Ref https://www.jooq.org/doc/latest/manual/code-generation/codegen-advanced/codegen-config-database/codegen-database-version-providers/#NA8349

Unfortunately, this solution only skips the codegen, not the container startup. Apparently jOOQ connects to the database before it checks with the `schemaVersionProvider`, even though the provider is actually a constant in this case. 

### Possible solution you'd like to see:

When using a constant `schemaVersionProvider`/`catalogVersionProvider`, it would be great if jOOQ were to compare the previously generated files earlier, and thus avoid initiating the JDBC connection if there are no changes and thus nothing to do.

### Possible workarounds:

None

### Versions:

- jOOQ: 3.14.15
- Java: 17
- Database (include vendor): Postgres 12
- OS: MacOS 12.2.1
- JDBC Driver (include name if inofficial driver): org.testcontainers.jdbc.ContainerDatabaseDriver

**Comment by lukaseder** [comment]
Thanks a lot for your suggestion. This had already been under consideration some time ago (though not sure if there's an issue). The open question is: How can the provider know what files to look at before it has queried the database for available schemas, knowing that even the `<inputSchema>` and `<inputCatalog>` configurations are optional, which leads to generating code for *all* available schemas?

**Comment by seanf** [comment]
> Thanks a lot for your suggestion. This had already been under consideration some time ago (though not sure if there's an issue). The open question is: How can the provider know what files to look at before it has queried the database for available schemas, knowing that even the `<inputSchema>` and `<inputCatalog>` configurations are optional, which leads to generating code for _all_ available schemas?

Yes, also FWIW I noticed that some of the generated files lack a `@Generated` annotation with the schema version. I definitely don't understand all the implications here, but would it be possible to output another file which contains more information about the inputs? Maybe that already happens, or it just doesn't matter. 

But in my case I would be happy to specify `<inputSchema>` and `<inputCatalog>` if it helped with incremental builds.

As a more general solution, perhaps it would help to have a higher level version provider, eg `org.jooq.meta.DatabaseVersionProvider` or `org.jooq.meta.MetaVersionProvider`, which versions the complete set of _all_ catalogs and schemas in the database. It wouldn't be appropriate in all cases, but I think it should  work when the database is directly generated from SQL files.

**Comment by lukaseder** [comment]
> As a more general solution, perhaps it would help to have a higher level version provider, eg `org.jooq.meta.DatabaseVersionProvider`

Yeah, you're right, that would work. It would also work well for those databases that consist only of a single schema, which are most of them, so the two types of version providers can be used interchangeably, except that this one can be used also to prevent checking for schema availability in the actual database.

A good name would be `GlobalVersionProvider`, as it acts like a global variable
